### PR TITLE
Added advanced Heavy Hitter Detection example

### PR DESCRIPTION
**Pull request by gnikol**
Heavy hitter tutorial used for CS344 in Stanford.

### CODE REVIEW

**Comment by jnfoster** [comment]
This is cool. 

We don't really keep code in the top-level directory. And it'd be nice to harmonize it with the build processes used for other tutorials. 

Would it be possible to make these changes before we merge?

**Comment by gnikol** [comment]
Sure. I'm not sure about where the tutorial should go that's why I put on the top directory waiting for suggestions. It obviously cannot go under the event-related directories so, should I put it under examples?

Regarding the building, my code uses p4app, which I think is also used by some of the other tutorials. Could you be more specific about what I should do?

**Comment by gnikol** [comment]
- I added the same same build process like in SIGCOMM 2017 (e.g. HULA).
- Fixed the skeleton program
- Added comments in the TODO sections
- Updated README

File: Advanced_HHD/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Instructions
+
+You have to fill in the parts of simple_router.p4 marked with "TODO". You can find a solution in the solution folder.
```
[the remark this answers is not in the document]
Comment (gnikol) [reply Advanced_HHD/README.md]: Done

File: Teaching/Stanford_CS344_2018/README.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+3. Type `exit` to leave each Mininet command line.
+
+### A note about the control plane
+
+A P4 program defines a packet-processing pipeline, but the rules
+within each table are inserted by the control plane. When a rule
+matches a packet, its action is invoked with parameters supplied by
+the control plane as part of the rule.
+
+In this exercise, we have already implemented the the control plane
```
Comment (antoninbas) [inline Teaching/Stanford_CS344_2018/README.md]: typo

File: Teaching/Stanford_CS344_2018/README.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+times for small, short-lived flows. Detecting heavy hitters allows us to treat them 
+differently, e.g. we can put their packets in low priority queues, allowing
+packets of other flows to face little or no congestion.
+
+In this example, you will implement a heavy hitter detection filter within 
+a router. You can find a skeleton of the program in simple_router.p4. In that
+file, you have to fill in the parts that are marked with TODO.
+
+This example is based on [count-min sketch](http://theory.stanford.edu/~tim/s15/l/l2.pdf).
+In fact, we use two count-mi sketches which are reset with an offset
```
Comment (antoninbas) [inline Teaching/Stanford_CS344_2018/README.md]: typo

File: Teaching/Stanford_CS344_2018/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+logic for you. As part of invoking `run.sh`, a set of rules is generated
+by `setup.py` and when bringing up the Mininet instance, these
+packet-processing rules are installed in the tables of
+the switch. These are defined in the `simple_router.config` file.
+
+## Step 2: Implement the heavy hitter detection filter
+
+The `simple_router.p4` file contains a skeleton P4 program with key pieces of
+logic replaced by `TODO` comments. Your implementation should follow
+the structure given in this file---replace each `TODO` with logic
```
Comment (antoninbas) [inline Teaching/Stanford_CS344_2018/README.md]: does the `---` mean something?

File: Teaching/utils/mininet/appcontroller.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        entries = {}
+        for sw in self.topo.switches():
+            entries[sw] = []
+            if 'switches' in self.conf and sw in self.conf['switches'] and 'entries' in self.conf['switches'][sw]:
+                extra_entries = self.conf['switches'][sw]['entries']
+                if type(extra_entries) == list: # array of entries
+                    entries[sw] += extra_entries
+                else: # path to file that contains entries
+                    entries[sw] += self.read_entries(extra_entries)
+            #entries[sw] += [
```
Comment (antoninbas) [inline Teaching/utils/mininet/appcontroller.py]: I would recommend just removing what you don't need (it seems that these are read from the file)

File: Teaching/Stanford_CS344_2018/simple_router.config.template [hunk 6]
```diff
@@ -0,0 +1,12 @@
+set_crc16_parameters calc_2 0x1021 0xffff 0x0000 false false
```
Comment (antoninbas) [inline Teaching/Stanford_CS344_2018/simple_router.config.template]: I'm not too familiar with p4app, is there a reason for the `.template` suffix in the file name?

File: Teaching/Stanford_CS344_2018/README.md (hunk 2 above)
Comment (gnikol) [reply Teaching/Stanford_CS344_2018/README.md]: Fixed

File: Teaching/Stanford_CS344_2018/README.md (hunk 3 above)
Comment (gnikol) [reply Teaching/Stanford_CS344_2018/README.md]: Fixed

File: Teaching/Stanford_CS344_2018/README.md (hunk 4 above)
Comment (gnikol) [reply Teaching/Stanford_CS344_2018/README.md]: Fixed

File: Teaching/utils/mininet/appcontroller.py (hunk 5 above)
Comment (gnikol) [reply Teaching/utils/mininet/appcontroller.py]: Fixed

File: Teaching/Stanford_CS344_2018/simple_router.config.template (hunk 6 above)
Comment (gnikol) [reply Teaching/Stanford_CS344_2018/simple_router.config.template]: p4app needs simple_router.config which generated by setup.py by appending a set of rules to the contents of simple_router.config.template.

**Comment by gnikol** [comment]
@jnfoster any more input, or we can proceed with merging?

**Pull request merged**